### ISSUE TITLE
The CI docs skip hack is fundamentally broken WRT branch protection settings

### ISSUE THREAD

**Issue by geekosaur** [haskell/cabal#10365]
**Describe the bug**
See https://github.com/haskell/cabal/actions/runs/10888849265/job/30214209267.
TL;DR: the "validate skip" post job succeeded immediately, so branch protection considered "Validate post job" to have succeeded and the PR was merged as soon as the other required jobs completed.

It's not clear whether "Bootstrap post job" has the same problem: the post job also completed¸ but so also did all the actual bootstrap jobs. It's possible that the so-called "quick jobs" take long enough for it to finish (I regularly see that getting canceled when I push a fix to something revealed within the first few minutes of a PR build). Or some other required job; I'd have to remind myself what's actually in the branch protection requirements to make more than a slightly-informed guess.

This doesn't seem to happen on "pull_request" or "push" jobs. I suspect that's because branch protection doesn't come into play on those, suggesting that it's overactive branch protection checks that are the real problem (which is a GitHub issue).

**Comment by geekosaur** [comment]
Okay, it's a little more complicated than that. The post job works as expected for `pull_request` and `push`, just not for the run before it's merged for some reason. I _think_ the problem might instead be that Mergify takes over the branch protection rules, and doesn't evaluate them the same way GitHub Actions does.

**Comment by geekosaur** [comment]
Interestingly, https://github.com/haskell/cabal/pull/10357 ran the full set. But I had noticed that the marathon run of backports went in unexpectedly quickly, so now I need to check how the release branch rules differ from `master`. (I have worked on those, but only on stuff _before_ it starts the merge.)

**Comment by geekosaur** [comment]
The behavior is fairly inconsistent. I still think this is something about how Mergify evaluates the branch protection rules (which are the same between `3.*` and `master` modulo fourmolu and hlint) and want to test whether removing `Validate post job` and `Bootstrap post job` from the required list and doing it in Mergify config will work.

That said, I also wonder why `Validate skip` is even running: supposedly it only triggers on docs, as configured. Supporting this is that it doesn't even show up in the CI summary. But there's still that `Validate post job` clearly was running `exit 0` (from `Validate skip` instead of the outcomes check in `Validate`.

**Comment by geekosaur** [comment]
Okay, looked over triggers for `Validate skip`. As written,

- it runs only for PRs targeting `master`
- it runs on any PR that touches documentation, even if it also touches code

So if we have a PR targeting `3.12`/`3.14` or that doesn't affect docs, the collision doesn't happen and CI runs fully. If it does affect docs, `Validate skip` finishes first and Mergify's interpretation of the branch protection rules is satisfied. I'm going to see if there's to make this "either/or" with `Validate`, and allow it to run on release branches as well. (Currently I think we get away with its configuration because both `push` (constrained) and `pull_request` (not constrained) are true, so it runs on release branches.)

**Comment by geekosaur** [comment]
Okay, so GHA controls aren't flexible enough. But Mergify can _I think_ be told to accept "skipped" as "success" — or may even do so by default, unlike GHA. So we'd again just move that branch protection rule to Mergify config.

**Comment by geekosaur** [comment]
Another observation: our CI proceeds in weird steps that may be contributing here. For example, our validate post job checks all the preceding steps, but _those steps don't exist_ as far as both GitHub Actions and Mergify are concerned while earlier steps are running; they're only populated when earlier steps are finished, for some reason.

That said, I have seen cases where the initial validate job is still running when something is merged, so this probably isn't (usually?) it. But I still wonder why those later jobs aren't listed as pending.

**Comment by geekosaur** [comment]
It's worse than I thought… and it's not Mergify that's at fault, it is definitely GitHub.
```
[06 14:52:06] <haskellbridge> <geekosaur> caught one in the act! and it looks like it's something with github
[06 14:54:33] <haskellbridge> <geekosaur> https://github.com/haskell/cabal/pull/10417 just got merged while CI was still running. it required a rebase beforehand, which happened as expected. GitHub showed that the PR wasn't mergeable for several minutes, then cleared that and Mergify immediately merged it. Notable: "Validate post job" shows as complete even though all the validate jobs are still running
[06 14:54:53] <haskellbridge> <geekosaur> (likewise with "Bootstrap post job")
[06 14:56:01] <haskellbridge> <geekosaur> _but_ this can't be the skip jobs _per se_ because they completed immediately.
[06 14:57:07] <haskellbridge> <geekosaur> …Mergify thinks it's still not merged!
[06 14:57:35] <haskellbridge> <geekosaur> I think it merged as soon as the so-called "quick jobs" finished
[06 15:02:21] <haskellbridge> <geekosaur> so. one of my working hypotheses was that GitHub and Mergify disagreed about the thing we've been relying on for the skip hack. But it was GitHub that called the job finished, not Mergify, so the whole basis for the skip hack is incorrect. GitHub (again, _not_ Mergify) considered the post job complete as soon as the skip job's version ran. This doesn't happen during normal PR checks because they're not checked...
[06 15:02:26] <haskellbridge> ... against the branch protection rules.
[06 15:04:25] <haskellbridge> <geekosaur> So now my top question is: why did GitHub auto-merge the PR insytead of waiting for Mergify to complete its checks?
```

**Comment by geekosaur** [comment]
```
[06 15:18:49] <haskellbridge> <geekosaur> hm, but github claims mergify committed it, while mergify says "The pull request embarked with master (b1e842d) will be merged soon"
[06 15:19:33] <haskellbridge> <geekosaur> mergify dashboard says nothing running…
[06 15:20:12] <haskellbridge> <geekosaur> methinks we have here a major disconnect between mergify and github, and both are complicit
```

**Comment by ulysses4ever** [comment]
I again propose to revert the hack and stop wasting the precious resource for investigating it any further.

Mergify can't be not a part of the equation because the hack is known to work for pure GitHub workflows?

**Comment by geekosaur** [comment]
As far as I can determine, it's branch protection that's failing, not Mergify; Mergify is just obeying the branch protection rules, and the sequence of events when the PR was committed clearly indicate that it was GitHub, not Mergify, that went green on branch protection inappropriately. Possibly the dual intention of our post jobs (both reducing the checks in e.g. validate to a single mega-check and enabling the docs skip hack) is contributing, because branch protection no longer knows about the details?

In any case, I now think #10431 will fix it because Mergify would no longer jump the gun when GitHub's branch protection does.

**Comment by geekosaur** [comment]
Also, I think reverting the docs hack is itself a hackaround, because this will still be lying in wait to bite us in some other way.

**Comment by geekosaur** [comment]
I reported Mergify's jumping the gun on the checks as a bug; sadly, the only feedback I got was a "thank you" from the bug reporting page, not even a bug number.

**Comment by ulysses4ever** [comment]
Fair enough.

> https://github.com/haskell/cabal/pull/10431 will fix it

🤞

> reported Mergify's jumping the gun on the checks as a bug

Is there a public link?

**Comment by geekosaur** [comment]
Like I said, no feedback whatsoever aside from the "thank you". I had intended to put a link here, but apparently their bugs are not public.

**Comment by geekosaur** [comment]
The Mergify tweak didn't work; it again merged while CI was running. Mergify support asked for more information which I gave them, and I'm waiting to hear back.
